### PR TITLE
rgw: User Policy

### PR DESCRIPTION
**Pull request by pritha-srivastava**
This PR:
1. Adds code for User Policy evaluation for Users/ buckets and objects.
2. Adds REST APIs for User Policy Manipulation.
3. Adds code for user permission evaluation based on User Policies for REST APIs for Roles.

### CODE REVIEW

**Comment by votdev** [comment]
Can you please add a description what this PR is for or point to a document? Maybe this is something we need to add to the mgr dashboard.

File: src/rgw/rgw_rest_user_policy.cc [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    if (op_ret < 0) {
+      op_ret = -ERR_INTERNAL_ERROR;
+    }
+  } catch (rgw::IAM::PolicyParseException& e) {
+    ldout(s->cct, 20) << "failed to parse policy: " << e.what() << dendl;
+    op_ret = -ERR_MALFORMED_DOC;
+  }
+  ldout(s->cct, 20) << "op_ret is : " << op_ret << dendl;
+}
+
```
Comment (adamemerson) [inline src/rgw/rgw_rest_user_policy.cc]: Looks reasonable

File: src/rgw/rgw_iam_policy.h [hunk 2]
```diff
@@ -40,77 +41,88 @@ struct rgw_bucket;
...
+static constexpr std::uint64_t iamGetRolePolicy = 65;
+static constexpr std::uint64_t iamListRolePolicies = 66;
+static constexpr std::uint64_t iamDeleteRolePolicy = 67;
+
+static constexpr std::uint64_t s3Count = s3DeleteObjectVersionTagging + 1;
+static constexpr std::uint64_t allCount = iamDeleteRolePolicy + 1;
+
+using Action_t = bitset<allCount>;
+using NotAction_t = Action_t;
+
```
Comment (adamemerson) [inline src/rgw/rgw_iam_policy.h]: Excellent.

Comment (mattbenjamin) [reply src/rgw/rgw_iam_policy.h]: indeed

File: src/rgw/rgw_common.cc [hunk 3]
```diff
@@ -1061,12 +1061,23 @@ bool verify_user_permission(struct req_state * const s,
...
 
-  auto perm = op_to_perm(op);
+  if (op == rgw::IAM::s3CreateBucket || op == rgw::IAM::s3ListAllMyBuckets) {
+    auto perm = op_to_perm(op);
 
-  return verify_user_permission_no_policy(s, user_acl, perm);
+    return verify_user_permission_no_policy(s, user_acl, perm);
+  }
+
+  if (usr_policy_res == Effect::Pass) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_common.cc]: ack

File: src/rgw/rgw_rest_role.h [hunk 4]
```diff
@@ -11,10 +13,11 @@ class RGWRestRole : public RGWRESTOp {
   string policy_name;
   string perm_policy;
   string path_prefix;
-
+  RGWRole _role;
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest_role.h]: why leading underscore?  generally, would avoid that (cosmetic)

File: src/rgw/rgw_common.cc [hunk 5]
```diff
@@ -1023,7 +1023,53 @@ string RGWHTTPArgs::sys_get(const string& name, bool * const exists) const
   return e ? iter->second : string();
 }
 
+namespace {
```
Comment (mattbenjamin) [inline src/rgw/rgw_common.cc]: why anonymous namespace?

File: src/test/rgw/test_rgw_iam_policy.cc [hunk 6]
```diff
@@ -546,6 +698,38 @@ string PolicyTest::example3 = R"(
...
+string PolicyTest::example6 = R"(
+{
+  "Version": "2012-10-17",
+  "Statement": {
+    "Effect": "Allow",
+    "Action": "*",
+    "Resource": "arn:aws:iam:::user/A"
+  }
+}
+)";
```
Comment (pritha-srivastava) [inline src/test/rgw/test_rgw_iam_policy.cc]: @adamemerson : I have added some unit tests, can you please take a look at them

**Comment by djgalloway** [comment]
jenkins test make check

File: src/test/rgw/test_rgw_iam_policy.cc (hunk 6 above)
Comment (pritha-srivastava) [reply src/test/rgw/test_rgw_iam_policy.cc]: @adamemerson : Can you please review the unit tests

**Comment by mattbenjamin** [comment]
http://pulpito.ceph.com/mbenjamin-2018-09-10_17:48:09-rgw-wip-user-policy-distro-basic-smithi/

**Pull request merged**